# Кинематографический анализ MovieLens

Кино — на удивление молодое искусство: чуть больше века против тысячелетий устной традиции. По сути это прямое продолжение одного из древнейших человеческих занятий — рассказывать истории у костра, вместо которого теперь экран.

Новизна нашего времени не в самих историях, а в масштабе наблюдения за реакцией на них. Раньше о желаниях и мечтах общества судили по обрывочным свидетельствам — письмам, дневникам, кассовым сборам горстки театров. Сегодня миллионы людей сами, добровольно и в деталях, документируют свою реакцию на истории: ставят оценку, пишут тег, выбирают, что пересмотреть. MovieLens — один из таких архивов коллективной реакции.

В этом отчёте мы используем его, чтобы увидеть отражение — не только индустрии, но и нас самих: чего люди ищут в историях, от чего устают, чего с любопытством ждут и на что готовы тратить деньги. Разберём это в четырёх актах: эпохи и жанры, теги и ассоциации, оценки зрителей, экономика индустрии.

---

In [1]:
import movielens_analysis

%timeit movielens_analysis

51.2 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [2]:
DATA_DIR = "datasets/"

movies = movielens_analysis.Movies(DATA_DIR + "movies.csv")
tags = movielens_analysis.Tags(DATA_DIR + "tags.csv")
ratings = movielens_analysis.Ratings(DATA_DIR + "ratings.csv", DATA_DIR + "movies.csv")
links = movielens_analysis.Links(DATA_DIR + "links.csv", DATA_DIR + "movies.csv")
dvis = movielens_analysis.DataViz()

%timeit movies, tags, ratings, links, dvis

94.7 ns ± 1.12 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


---

# Акт I: Эпохи и жанры

Жанр — это ответ индустрии на вопрос «что тебе сейчас нужно почувствовать?». Прежде чем смотреть на оценки и деньги, разберёмся с этим фундаментом: когда и что именно снимали, и что выбор жанров говорит о коллективных потребностях аудитории.

### ❓ Вопрос 1: В какие годы мир увидел больше всего премьер?

In [3]:
dvis.bar_chart(
    movies.dist_by_release(),
    title="Movie Distribution by year",
    max_bar_width=30,
)

%timeit movies.dist_by_release()

─────────────────────────────────────────────────
  Movie Distribution by year
─────────────────────────────────────────────────
  2002 |██████████████████████████████|   311 ( 3.2%)
  2006 |████████████████████████████  |   295 ( 3.0%)
  2001 |████████████████████████████  |   294 ( 3.0%)
  2007 |███████████████████████████   |   284 ( 2.9%)
  2000 |███████████████████████████   |   283 ( 2.9%)
  2009 |███████████████████████████   |   282 ( 2.9%)
  2003 |██████████████████████████    |   279 ( 2.9%)
  2004 |██████████████████████████    |   279 ( 2.9%)
  2014 |██████████████████████████    |   278 ( 2.9%)
  1996 |██████████████████████████    |   276 ( 2.8%)
  2015 |██████████████████████████    |   274 ( 2.8%)
  2005 |██████████████████████████    |   273 ( 2.8%)
  2008 |█████████████████████████     |   269 ( 2.8%)
  1999 |█████████████████████████     |   263 ( 2.7%)
  1997 |█████████████████████████     |   260 ( 2.7%)
  1995 |████████████████████████      |   259 ( 2.7%)
  1998 

> Никаких девяностых: заветный пик — **2002 год, 311 премьер**, почти вровень со всей серединой нулевых. Хотя расцвет видеопроката принято относить к 90-м, реальный вал релизов пришёлся на цифровую эпоху: удешевление съёмки и монтажа открыло дорогу независимому кино.

Если посмотреть на общую картину, окажется, что подавляющему большинству от историй нужно всего два состояния: порадоваться и посмеяться — или прочувствовать и порефлексировать.

### ❓ Вопрос 2: Какие жанры были популярны на протяжении всей истории кино?

In [46]:
dvis.bar_chart(
    movies.dist_by_genres(),
    title="Genres by Frequency",
    max_bar_width=30,
)

%timeit movies.dist_by_genres()

────────────────────────────────────────────────────────
  Genres by Frequency
────────────────────────────────────────────────────────
        Drama |██████████████████████████████|  4361 (19.8%)
       Comedy |█████████████████████████     |  3756 (17.0%)
     Thriller |█████████████                 |  1894 ( 8.6%)
       Action |████████████                  |  1828 ( 8.3%)
      Romance |██████████                    |  1596 ( 7.2%)
    Adventure |████████                      |  1263 ( 5.7%)
        Crime |████████                      |  1199 ( 5.4%)
       Sci-Fi |██████                        |   980 ( 4.4%)
       Horror |██████                        |   978 ( 4.4%)
      Fantasy |█████                         |   779 ( 3.5%)
     Children |████                          |   664 ( 3.0%)
    Animation |████                          |   611 ( 2.8%)
      Mystery |███                           |   573 ( 2.6%)
  Documentary |███                           |   440 ( 2.0%)
          

> Итог предсказуем и в то же время красноречив: драма и комедия вместе — это **почти 37% всего каталога**. Это не значит, что зрители не любят фантастику или хоррор — просто два этих жанра оказались самыми дешёвыми в производстве и самыми универсальными по спросу, поэтому индустрия штампует их из десятилетия в десятилетие.

Иногда привычных состояний — посмеяться или порефлексировать — недостаточно. Временами хочется чего-то необузданного, что не укладывается ни в одну привычную категорию.

### ❓ Вопрос 3: Какие фильмы объединили в себе рекордное число жанров сразу?

In [ ]:
dvis.bar_chart(
    movies.most_genres(5),
    title="Top 5 Movies by Genre Count",
    max_bar_width=25,
)

%timeit movies.most_genres(5)

────────────────────────────────────────────────────────────────────────────────
  Top 5 Movies by Genre Count
────────────────────────────────────────────────────────────────────────────────
                             Rubber (2010) |█████████████████████████|    10 (25.6%)
  Patlabor: The Movie (Kidô keisatsu pator |████████████████████     |     8 (20.5%)
                              Mulan (1998) |█████████████████        |     7 (17.9%)
           Who Framed Roger Rabbit? (1988) |█████████████████        |     7 (17.9%)
                      Osmosis Jones (2001) |█████████████████        |     7 (17.9%)
────────────────────────────────────────────────────────────────────────────────
  Total: 39  |  Unique: 5

12 ms ± 355 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


> *Rubber* (2010, в русском прокате — «Шина») — фильм про шину-убийцу, обретшую сознание, — единственный в выборке собрал сразу 10 жанровых ярлыков. Формально это можно списать на то, что классификаторам не хватило существующих категорий. Но по сути это и есть выход за рамки привычного: усталость от знакомых форм иногда ищет разрядку в намеренно абсурдном и неклассифицируемом.

Если жанровый хаос — редкое исключение, то большинство фильмов, наоборот, комбинируют жанры по проверенным формулам.

### 💡 Бонусный вопрос 4: Какие жанровые группы стали золотым стандартом киноиндустрии?

In [5]:
dvis.bar_chart(
    movies.genre_combinations(10),
    title="Top 10 Genre Combinations",
    max_bar_width=30,
)

%timeit movies.genre_combinations(10)

─────────────────────────────────────────────────────────────────
  Top 10 Genre Combinations
─────────────────────────────────────────────────────────────────
                 Drama |██████████████████████████████|  1053 (24.9%)
                Comedy |██████████████████████████    |   946 (22.4%)
          Comedy|Drama |████████████                  |   435 (10.3%)
        Comedy|Romance |██████████                    |   363 ( 8.6%)
         Drama|Romance |█████████                     |   349 ( 8.2%)
           Documentary |█████████                     |   339 ( 8.0%)
  Comedy|Drama|Romance |███████                       |   276 ( 6.5%)
        Drama|Thriller |████                          |   168 ( 4.0%)
                Horror |████                          |   167 ( 3.9%)
       Horror|Thriller |███                           |   135 ( 3.2%)
─────────────────────────────────────────────────────────────────
  Total: 4,231  |  Unique: 10

4.81 ms ± 179 μs per loop (mean ± std. dev.

> Оказывается, «чистые» жанры без примесей популярнее любых гибридов: фильмы, помеченные просто как `Drama` или просто как `Comedy`, обходят даже привычную формулу romcom `Comedy|Romance`. Похоже, публика ценит фильм, который точно знает, чем он хочет быть.

---

# Акт II: Теги и ассоциации

Жанры — это официальная классификация, придуманная индустрией. Теги — куда менее формальный слой: то первое слово, которое зритель сам, без всякой инструкции, счёл нужным оставить после титров. Жанры показывают, что индустрия производит, теги говорят, что из этого действительно откладывается у людей в голове.

### ❓ Вопрос 5: Какие теги всплывают чаще всего?

In [6]:
dvis.bar_chart(
    tags.most_popular(10),
    title="Top 10 Most Popular Tags",
    max_bar_width=30,
)

%timeit tags.most_popular(10)

──────────────────────────────────────────────────────────────
  Top 10 Most Popular Tags
──────────────────────────────────────────────────────────────
   In Netflix queue |██████████████████████████████|   131 (37.6%)
        atmospheric |████████                      |    36 (10.3%)
          superhero |█████                         |    24 ( 6.9%)
  thought-provoking |█████                         |    24 ( 6.9%)
              funny |█████                         |    23 ( 6.6%)
             Disney |█████                         |    23 ( 6.6%)
            surreal |█████                         |    23 ( 6.6%)
           religion |█████                         |    22 ( 6.3%)
             sci-fi |████                          |    21 ( 6.0%)
        dark comedy |████                          |    21 ( 6.0%)
──────────────────────────────────────────────────────────────
  Total: 348  |  Unique: 10

571 μs ± 17.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> Самый популярный тег — вообще не про кино: это служебная пометка `In Netflix queue` (131 раз) — люди тегировали не эмоцию, а напоминалку самим себе. Зато сразу за ней идут куда более человечные слова — `atmospheric` и `thought-provoking`. Похоже, зрителям MovieLens важнее атмосфера и пища для ума, чем спецэффекты.

Иногда одного слова не хватает — и рождается мини-рецензия прямо в поле для тега.

### ❓ Вопрос 6: Какие теги имеют наибольшее число слов и символов?

In [7]:
dvis.bar_chart(
    tags.most_words(10),
    title="Top 10 Tags by Word Count",
    max_bar_width=30,
)

%timeit tags.most_words(10)

─────────────────────────────────────────────────────────────────────────────────────
  Top 10 Tags by Word Count
─────────────────────────────────────────────────────────────────────────────────────
  Something for everyone in this one... sa |██████████████████████████████|    16 (21.6%)
  the catholic church is the most corrupt  |██████████████████            |    10 (13.5%)
  villain nonexistent or not needed for go |███████████████               |     8 (10.8%)
      It was melodramatic and kind of dumb |█████████████                 |     7 ( 9.5%)
  06 Oscar Nominated Best Movie - Animatio |███████████                   |     6 ( 8.1%)
  stop using useless characters for filler |███████████                   |     6 ( 8.1%)
  r:disturbing violent content including r |███████████                   |     6 ( 8.1%)
                        based on a TV show |█████████                     |     5 ( 6.8%)
                   stop looking at me swan |█████████                     |     

In [15]:
display(tags.longest(10))

%timeit tags.longest(10)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 Oscar Nominated Best Movie - Animation',
 'stop using useless characters for filler',
 'Academy award (Best Supporting Actress)',
 'Oscar (Best Effects - Visual Effects)',
 'audience intelligence underestimated',
 'It was melodramatic and kind of dumb']

548 μs ± 26.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> Рекордсмен по многословности — признание длиной в целое предложение: *«Something for everyone in this one... saw it without and plan on seeing it with kids!»* — шестнадцать слов чистой родительской рефлексии вместо тега.

А теперь взглянем на пересечение: какие теги одновременно состоят из множества слов и имеют рекордную длину в символах?

In [13]:
display(tags.most_words_and_longest(5))

%timeit tags.most_words_and_longest(5)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 '06 Oscar Nominated Best Movie - Animation']

3.57 ms ± 243 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Некоторые теги — это прямое указание на культурный ориентир: студию, франшизу, целый пласт воображения. Показательный пример ниже — поиск по слову «robots»: он поднимает истории про андроидов и механических существ, то самое любопытство к будущему, что мы уже видели в жанровой статистике (Sci-Fi, Fantasy).

### ❓ Вопрос 7: Как зрители описывают знаковые вселенные и студии (например, 'pixar')?

In [16]:
display(tags.tags_with("pixar"))

%timeit tags.tags_with("pixar")

['Pixar', 'pixar']

1.78 ms ± 51.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [17]:
display(tags.tags_with("robots"))

%timeit tags.tags_with("robots")

['ROBOTS AND ANDROIDS', 'robots']

1.48 ms ± 44 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


> Оба найденных тега — `ROBOTS AND ANDROIDS` и `robots` — подтверждают то, что уже видно в жанровой статистике: истории о будущем, ИИ и механических существах остаются устойчивым источником сюжетов на протяжении всей истории каталога, а не проходящей модой одного десятилетия.

Одни зрители ограничиваются словом-меткой, другие превращают тег в маленькое высказывание. Кто из них самый изобретательный?

### 💡 Бонусный вопрос 8: Кто из пользователей обладает самым богатым и неординарным вокабуляром?

In [21]:
dvis.bar_chart(
    tags.user_vocabulary_richness(10),
    title="Top 10 Users by Richness",
    max_bar_width=30,
)

%timeit tags.user_vocabulary_richness(10)

─────────────────────────────────────────────────
  Top 10 Users by Richness
─────────────────────────────────────────────────
     2 |██████████████████████████████|   1.0 (10.3%)
   305 |██████████████████████████████|   1.0 (10.3%)
   336 |██████████████████████████████|   1.0 (10.3%)
   356 |██████████████████████████████|   1.0 (10.3%)
   419 |██████████████████████████████|   1.0 (10.3%)
   462 |██████████████████████████████|   1.0 (10.3%)
   606 |██████████████████████████████|   1.0 (10.3%)
   318 |███████████████████████████   |  0.93 ( 9.6%)
   357 |███████████████████████████   |  0.93 ( 9.6%)
   119 |█████████████████████████     |  0.86 ( 8.8%)
─────────────────────────────────────────────────
  Total: 9.72  |  Unique: 10

1.26 ms ± 78.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> Здесь статистическая ловушка: у десятков пользователей «идеальная» richness = 1.0. На деле это означает не богатство словаря, а то, что человек поставил всего один тег за всю историю — при выборке из одного элемента показатель разнообразия математически не может быть иным. Хороший повод помнить, что красивая метрика не всегда означает то, что кажется на первый взгляд.

---

# Акт III: Взгляд сквозь оценки

Жанры и теги показывают, что люди выбирают и как об этом говорят. Оценки — это уже более весомая реакция: как только что рассказанная история была воспринята. Здесь интересно смотреть на закономерности поведения аудитории.

### ❓ Вопрос 9: Как менялась активность выставления оценок с годами и как распределились баллы?

In [22]:
dvis.distribution_by_key(
    ratings.movies.dist_by_year(),
    title="Ratings Activity by Year",
    max_bar_width=40,
)

%timeit ratings.movies.dist_by_year()

───────────────────────────────────────────────────────────
  Ratings Activity by Year
───────────────────────────────────────────────────────────
  1996 |████████████████████████                |  6040 ( 6.0%)
  1997 |███████                                 |  1916 ( 1.9%)
  1998 |██                                      |   507 ( 0.5%)
  1999 |█████████                               |  2439 ( 2.4%)
  2000 |████████████████████████████████████████| 10061 (10.0%)
  2001 |███████████████                         |  3922 ( 3.9%)
  2002 |█████████████                           |  3478 ( 3.4%)
  2003 |███████████████                         |  4014 ( 4.0%)
  2004 |█████████████                           |  3279 ( 3.3%)
  2005 |███████████████████████                 |  5813 ( 5.8%)
  2006 |████████████████                        |  4059 ( 4.0%)
  2007 |████████████████████████████            |  7114 ( 7.1%)
  2008 |█████████████████                       |  4351 ( 4.3%)
  2009 |█████████████

> Пик активности пришёлся на 2000 год (10 061 оценка) — но это, скорее, не бум кинопроизводства, а бум самого MovieLens: именно тогда сервис начал стремительно набирать пользователей, и они бросились переоценивать всё, что смотрели раньше.

In [23]:
dvis.distribution_by_key(
    ratings.movies.dist_by_rating(),
    title="Rating Distribution",
    max_bar_width=40,
)

%timeit ratings.movies.dist_by_rating()

───────────────────────────────────────────────────────────
  Rating Distribution
───────────────────────────────────────────────────────────
   0.5 |██                                      |  1370 ( 1.4%)
   1.0 |████                                    |  2811 ( 2.8%)
   1.5 |██                                      |  1791 ( 1.8%)
   2.0 |███████████                             |  7551 ( 7.5%)
   2.5 |████████                                |  5550 ( 5.5%)
   3.0 |█████████████████████████████           | 20047 (19.9%)
   3.5 |███████████████████                     | 13136 (13.0%)
   4.0 |████████████████████████████████████████| 26818 (26.6%)
   4.5 |████████████                            |  8551 ( 8.5%)
   5.0 |███████████████████                     | 13211 (13.1%)
───────────────────────────────────────────────────────────
  Total: 100,836  |  Unique: 10

42.7 ms ± 1.88 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Мы — оптимисты: любимая оценка зрителей — **4.0** (26,8 тыс. раз), а вовсе не «твёрдая тройка». Похоже, на MovieLens если фильм досмотрели до конца, ставят не «нормально», а «хорошо».

Число оценок — грубый, но честный индикатор культурного веса фильма: сколько людей вообще сочли нужным включить его и досмотреть до конца.

### ❓ Вопрос 10: Какие фильмы привлекли внимание рекордного числа зрителей?

In [24]:
dvis.bar_chart(
    ratings.movies.top_by_num_of_ratings(5),
    title="Top 5 Movies by Number of Ratings",
    max_bar_width=30,
)

%timeit ratings.movies.top_by_num_of_ratings(5)

─────────────────────────────────────────────────────────────────────────────
  Top 5 Movies by Number of Ratings
─────────────────────────────────────────────────────────────────────────────
               Forrest Gump (1994) |██████████████████████████████|   329 (21.8%)
  Shawshank Redemption, The (1994) |████████████████████████████  |   317 (21.0%)
               Pulp Fiction (1994) |███████████████████████████   |   307 (20.3%)
  Silence of the Lambs, The (1991) |█████████████████████████     |   279 (18.5%)
                Matrix, The (1999) |█████████████████████████     |   278 (18.4%)
─────────────────────────────────────────────────────────────────────────────
  Total: 1,510  |  Unique: 5

30.7 ms ± 426 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Три титана 1994 года держат пьедестал: *Forrest Gump*, *The Shawshank Redemption* и *Pulp Fiction* — фильмы, которые проверяют время лучше любых блокбастеров своего десятилетия.

### ❓ Вопрос 11: Кто возглавляет рейтинг признания по среднему и медианному баллу?

In [25]:
dvis.bar_chart(
    ratings.movies.top_by_ratings(5, metric="average"),
    title="Top 5 by Average Rating",
    max_bar_width=30,
)

%timeit ratings.movies.top_by_ratings(5, metric="average")

─────────────────────────────────────────────────────────────────────────────────────
  Top 5 by Average Rating
─────────────────────────────────────────────────────────────────────────────────────
  The Jinx: The Life and Deaths of Robert  |██████████████████████████████|   5.0 (20.0%)
           Galaxy of Terror (Quest) (1981) |██████████████████████████████|   5.0 (20.0%)
                Alien Contamination (1980) |██████████████████████████████|   5.0 (20.0%)
            I'm the One That I Want (2000) |██████████████████████████████|   5.0 (20.0%)
                       Lesson Faust (1994) |██████████████████████████████|   5.0 (20.0%)
─────────────────────────────────────────────────────────────────────────────────────
  Total: 25.0  |  Unique: 5

69.1 ms ± 5.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [26]:
dvis.bar_chart(
    ratings.movies.top_by_ratings(5, metric="median"),
    title="Top 5 by Median Rating",
    max_bar_width=30,
)

%timeit ratings.movies.top_by_ratings(5, metric="median")

─────────────────────────────────────────────────────────────────────────────────────
  Top 5 by Median Rating
─────────────────────────────────────────────────────────────────────────────────────
  The Jinx: The Life and Deaths of Robert  |██████████████████████████████|   5.0 (20.0%)
           Galaxy of Terror (Quest) (1981) |██████████████████████████████|   5.0 (20.0%)
                Alien Contamination (1980) |██████████████████████████████|   5.0 (20.0%)
                            Troll 2 (1990) |██████████████████████████████|   5.0 (20.0%)
            I'm the One That I Want (2000) |██████████████████████████████|   5.0 (20.0%)
─────────────────────────────────────────────────────────────────────────────────────
  Total: 25.0  |  Unique: 5

75.3 ms ± 4.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Осторожно: идеальные пятёрки почти всегда означают, что фильм оценили буквально пару раз — это не «лучший фильм всех времён», а математический шум малой выборки. Настоящих фаворитов будем искать иначе — в следующем, бонусном вопросе про «скрытые шедевры».

Согласие в оценках отражает общий вкус, но самое интересное происходит там, где мнения раскалываются, — это обычно граница между разными эстетическими или мировоззренческими лагерями внутри одной аудитории.

### ❓ Вопрос 12: Какие фильмы вызвали максимальную дисперсию мнений?

In [27]:
dvis.bar_chart(
    ratings.movies.top_controversial(5),
    title="Most Controversial Movies (Variance)",
    max_bar_width=30,
)

%timeit ratings.movies.top_controversial(5)

─────────────────────────────────────────────────────────────────────────────────────
  Most Controversial Movies (Variance)
─────────────────────────────────────────────────────────────────────────────────────
  Ivan's Childhood (a.k.a. My Name is Ivan |██████████████████████████████| 10.12 (22.9%)
  Fanny and Alexander (Fanny och Alexander |██████████████████████████████| 10.12 (22.9%)
                             Lassie (1994) |███████████████████████       |   8.0 (18.1%)
               Zed & Two Noughts, A (1985) |███████████████████████       |   8.0 (18.1%)
                   Kwaidan (Kaidan) (1964) |███████████████████████       |   8.0 (18.1%)
─────────────────────────────────────────────────────────────────────────────────────
  Total: 44.239999999999995  |  Unique: 5

92.8 ms ± 1.54 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Самый глубокий раскол мнений вызывают не блокбастеры, а авторское кино — *«Иваново детство»* Тарковского и *«Фанни и Александр»* Бергмана. Зрители либо признают в них шедевр, либо категорически не принимают язык медленного, философского кино — равнодушных почти не остаётся.

Топ по количеству оценок — это про охват. Отдельный вопрос — какие фильмы не привлекли толпы, но заслужили стабильно высокую оценку тех немногих, кто их всё-таки нашёл.

### 💡 Бонусный вопрос 13: Где скрываются «неизвестные шедевры» с высочайшим баллом, но умеренным числом оценок?

In [28]:
dvis.bar_chart(
    ratings.movies.hidden_gems(5, min_ratings=15, max_ratings=50),
    title="Hidden Gems (High score, moderate count)",
    max_bar_width=30,
)

%timeit ratings.movies.hidden_gems(5, min_ratings=15, max_ratings=50)

─────────────────────────────────────────────────────────────────────────────────────
  Hidden Gems (High score, moderate count)
─────────────────────────────────────────────────────────────────────────────────────
          Streetcar Named Desire, A (1951) |██████████████████████████████|  4.47 (20.4%)
                                Ran (1985) |█████████████████████████████ |  4.43 (20.2%)
  Sunset Blvd. (a.k.a. Sunset Boulevard) ( |█████████████████████████████ |  4.33 (19.8%)
                       Hustler, The (1961) |█████████████████████████████ |  4.33 (19.8%)
                   Double Indemnity (1944) |████████████████████████████  |  4.32 (19.7%)
─────────────────────────────────────────────────────────────────────────────────────
  Total: 21.88  |  Unique: 5

49.3 ms ± 1.37 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Вот они, настоящие скрытые шедевры: *A Streetcar Named Desire* и *Sunset Blvd.* — фильмы полувековой давности, которые никогда не попадут в топ по числу оценок, но у тех немногих, кто их посмотрел, стабильно выбивают 4.3+ балла. Возможно, лучший фильм на вечер — вовсе не тот, что все сейчас обсуждают.

От фильмов — к самой аудитории. Кто эти люди, что скрываются за строчками датасета?

### ❓ Вопрос 14: А как ведут себя сами зрители? Кто ставит оценки чаще всего и кто из них наиболее непредсказуем?

In [ ]:
dvis.bar_chart(
    ratings.users.dist_by_num_of_ratings(),
    title="Users by Number of Ratings Given",
    max_bar_width=30,
)

%timeit ratings.users.dist_by_num_of_ratings()

─────────────────────────────────────────────────
  Users by Number of Ratings Given
─────────────────────────────────────────────────
   414 |██████████████████████████████|  2698 ( 2.7%)
   599 |███████████████████████████   |  2478 ( 2.5%)
   474 |███████████████████████       |  2108 ( 2.1%)
   448 |████████████████████          |  1864 ( 1.8%)
   274 |██████████████                |  1346 ( 1.3%)
   610 |██████████████                |  1302 ( 1.3%)
    68 |██████████████                |  1260 ( 1.2%)
   380 |█████████████                 |  1218 ( 1.2%)
   606 |████████████                  |  1115 ( 1.1%)
   288 |███████████                   |  1055 ( 1.0%)
   249 |███████████                   |  1046 ( 1.0%)
   387 |███████████                   |  1027 ( 1.0%)
   182 |██████████                    |   977 ( 1.0%)
   307 |██████████                    |   975 ( 1.0%)
   603 |██████████                    |   943 ( 0.9%)
   298 |██████████                    |   939 ( 0.9%)
 

In [30]:
dvis.bar_chart(
    ratings.users.dist_by_rating(metric="average"),
    title="Users by Their Average Given Rating",
    max_bar_width=30,
)

%timeit ratings.users.dist_by_rating(metric="average")

─────────────────────────────────────────────────
  Users by Their Average Given Rating
─────────────────────────────────────────────────
    53 |██████████████████████████████|   5.0 ( 0.2%)
   251 |█████████████████████████████ |  4.87 ( 0.2%)
   515 |█████████████████████████████ |  4.85 ( 0.2%)
    25 |████████████████████████████  |  4.81 ( 0.2%)
    30 |████████████████████████████  |  4.74 ( 0.2%)
   523 |████████████████████████████  |  4.69 ( 0.2%)
   348 |████████████████████████████  |  4.67 ( 0.2%)
   171 |███████████████████████████   |  4.63 ( 0.2%)
   452 |███████████████████████████   |  4.56 ( 0.2%)
    43 |███████████████████████████   |  4.55 ( 0.2%)
   122 |███████████████████████████   |  4.55 ( 0.2%)
   371 |███████████████████████████   |  4.55 ( 0.2%)
   441 |███████████████████████████   |  4.52 ( 0.2%)
   400 |███████████████████████████   |  4.51 ( 0.2%)
    52 |██████████████████████████    |  4.48 ( 0.2%)
   538 |██████████████████████████    |  4.47 ( 0.2%

In [31]:
dvis.bar_chart(
    ratings.users.top_controversial(5),
    title="Most Unpredictable Users (Variance)",
    max_bar_width=30,
)

%timeit ratings.users.top_controversial(5)

─────────────────────────────────────────────────
  Most Unpredictable Users (Variance)
─────────────────────────────────────────────────
     3 |██████████████████████████████|  4.37 (25.8%)
    55 |██████████████████████        |  3.22 (19.0%)
   461 |██████████████████████        |  3.22 (19.0%)
   259 |████████████████████          |  3.05 (18.0%)
   329 |████████████████████          |  3.05 (18.0%)
─────────────────────────────────────────────────
  Total: 16.91  |  Unique: 5

59.5 ms ± 1.89 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Пользователь #414 поставил рекордные 2698 оценок — это уже не любитель кино, а исследователь-фанатик. А самым непредсказуемым оказался пользователь #3 — с разбросом мнений почти вдвое выше остальных: сегодня десятка, завтра единица без всякой видимой логики.

### 💡 Бонусный вопрос 15: Кто доминирует в киносообществе — строгие критики или благодарные фанаты?

In [32]:
dvis.bar_chart(
    ratings.users.critics_and_fans_ratio(),
    title="Audience Segmentation",
    max_bar_width=30,
)

%timeit ratings.users.critics_and_fans_ratio()

──────────────────────────────────────────────────────────
  Audience Segmentation
──────────────────────────────────────────────────────────
       moderate |██████████████████████████████|   418 (68.5%)
           fans |██████████                    |   143 (23.4%)
  harsh_critics |███                           |    49 ( 8.0%)
──────────────────────────────────────────────────────────
  Total: 610  |  Unique: 3

35.9 ms ± 246 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)


> Большинство зрителей — не судьи и не фанаты, а взвешенные наблюдатели: **68,5%** ставят умеренные оценки, и лишь **8%** — стабильно строгие критики. Если оценки — это отражение общества, то оно заметно снисходительнее к чужому труду, чем можно было бы ожидать.

---

# Акт IV: Режиссёры, бюджеты и окупаемость

Оценки и теги показывают, чего хотят зрители на словах. Более прямой индикатор — деньги: что люди реально готовы смотреть и оплачивать снова и снова. Подключаемся к TMDb за бюджетами и сборами и сравниваем это с тем, что уже увидели в реакции аудитории.

### ❓ Вопрос 16: Как получить полные метаданные фильма в реальном времени?

In [33]:
display(
    links.get_tmdb(
        ["1", "2", "3"], ["Director", "Budget", "Cumulative Worldwide Gross", "Runtime"]
    )
)
%timeit links.get_tmdb(["1", "2", "3"], ["Director", "Budget", "Cumulative Worldwide Gross", "Runtime"])

[[3, 'Howard Deutch', 25000000, 71518503, 101],
 [2, 'Joe Johnston', 65000000, 262821940, 104],
 [1, 'John Lasseter', 30000000, 962301978, 81]]

1.42 ms ± 64.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### ❓ Вопрос 17: Кто из режиссёров создал наибольшее число исследуемых шедевров?

In [37]:
dvis.bar_chart(
    links.top_directors(5),
    title="Top Directors by Movie Count",
    max_bar_width=30,
)

%timeit links.top_directors(5)

─────────────────────────────────────────────────────────────
  Top Directors by Movie Count
─────────────────────────────────────────────────────────────
      Joseph Ruben |██████████████████████████████|     1 (20.0%)
    Steve Oedekerk |██████████████████████████████|     1 (20.0%)
  Robert Rodriguez |██████████████████████████████|     1 (20.0%)
           Ang Lee |██████████████████████████████|     1 (20.0%)
   Martin Scorsese |██████████████████████████████|     1 (20.0%)
─────────────────────────────────────────────────────────────
  Total: 5  |  Unique: 5

1.53 ms ± 69.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> На этом крошечном срезе данных ни один режиссёр не повторяется дважды — выборка слишком мала, чтобы говорить о лидерах индустрии. Еще один повод помнить: не всякая красивая диаграмма означает статистически значимый вывод.

Здесь разворачивается главное несоответствие индустрии: одни фильмы сжигают бюджет и проваливаются, другие на скромные деньги делают кассу мечты.

### ❓ Вопрос 18: Какие картины потребовали колоссальных бюджетов и принесли рекордную чистую прибыль?

In [38]:
dvis.bar_chart(
    links.most_expensive(5),
    title="Most Expensive Movies (Budget)",
    max_bar_width=30,
)

%timeit links.most_expensive(5)

───────────────────────────────────────────────────────────────────────────
  Most Expensive Movies (Budget)
───────────────────────────────────────────────────────────────────────────
         Cutthroat Island (1995) |██████████████████████████████| 102000000 (28.6%)
              Money Train (1995) |████████████████████          | 68000000 (19.0%)
                  Jumanji (1995) |███████████████████           | 65000000 (18.2%)
  American President, The (1995) |██████████████████            | 62000000 (17.4%)
                GoldenEye (1995) |█████████████████             | 60000000 (16.8%)
───────────────────────────────────────────────────────────────────────────
  Total: 357,000,000  |  Unique: 5

1.36 ms ± 47.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [39]:
dvis.bar_chart(
    links.most_profitable(5),
    title="Most Profitable Movies (Gross - Budget)",
    max_bar_width=30,
)

%timeit links.most_profitable(5)

──────────────────────────────────────────────────────────────────────────────────
  Most Profitable Movies (Gross - Budget)
──────────────────────────────────────────────────────────────────────────────────
                       Toy Story (1995) |██████████████████████████████| 932301978 (53.8%)
                       GoldenEye (1995) |█████████                     | 292194034 (16.9%)
                         Jumanji (1995) |██████                        | 197821940 (11.4%)
  Ace Ventura: When Nature Calls (1995) |█████                         | 182385533 (10.5%)
                            Heat (1995) |████                          | 127400000 ( 7.4%)
──────────────────────────────────────────────────────────────────────────────────
  Total: 1,732,103,485  |  Unique: 5

1.61 ms ± 140 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> `Cutthroat Island` обошёлся в \$102 млн — колоссальную сумму по меркам 1995 года — и с треском провалился в прокате, едва не похоронив студию `Carolco`. А по-настоящему выстрелила скромная по бюджету (\$30 млн) `Toy Story`: чистая прибыль — \$932 млн. Урок стар как Голливуд: деньги на экране не гарантируют денег в кассе.

### ❓ Вопрос 19: Какие масштабные эпики требуют от зрителя наибольшего терпения?

In [41]:
dvis.bar_chart(
    links.longest(5),
    title="Longest Movies (Runtime in minutes)",
    max_bar_width=30,
)

%timeit links.longest(5)

─────────────────────────────────────────────────────────────────────────
  Longest Movies (Runtime in minutes)
─────────────────────────────────────────────────────────────────────────
                  Nixon (1995) |██████████████████████████████|   192 (23.8%)
                 Casino (1995) |███████████████████████████   |   179 (22.2%)
                   Heat (1995) |██████████████████████████    |   170 (21.1%)
  Sense and Sensibility (1995) |█████████████████████         |   136 (16.9%)
              GoldenEye (1995) |████████████████████          |   130 (16.1%)
─────────────────────────────────────────────────────────────────────────
  Total: 807  |  Unique: 5

1.43 ms ± 56 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> Три самых длинных фильма выборки — про власть, казино и погоню: *Nixon* (192 минуты), *Casino* (179) и *Heat* (170). Большим амбициям — большой хронометраж.

### ❓ Вопрос 20: Какова стоимость одной минуты киноаттракциона?

In [43]:
dvis.bar_chart(
    links.top_cost_per_minute(5),
    title="Cost Per Minute of Runtime",
    max_bar_width=30,
)

%timeit links.top_cost_per_minute(5)

───────────────────────────────────────────────────────────────────────────
  Cost Per Minute of Runtime
───────────────────────────────────────────────────────────────────────────
         Cutthroat Island (1995) |██████████████████████████████| 822580.65 (26.7%)
                  Jumanji (1995) |██████████████████████        | 625000.0 (20.3%)
              Money Train (1995) |██████████████████████        | 618181.82 (20.1%)
  American President, The (1995) |████████████████████          | 548672.57 (17.8%)
                GoldenEye (1995) |████████████████              | 461538.46 (15.0%)
───────────────────────────────────────────────────────────────────────────
  Total: 3,075,973.5  |  Unique: 5

1.57 ms ± 54.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> *Cutthroat Island* снова в антигероях: каждая минута фильма обошлась в **$822 580** — примерно как купить квартиру за один кадр. Деньги — не проблема.

Бюджет и прибыль — абсолютные цифры. Отдельный вопрос — эффективность каждого вложенного доллара.

### 💡 Бонусный вопрос 21: Какие проекты показали феноменальную окупаемость инвестиций (ROI)?

In [44]:
dvis.bar_chart(
    links.top_roi(5),
    title="Return on Investment (Gross / Budget)",
    max_bar_width=30,
)

%timeit links.top_roi(5)

──────────────────────────────────────────────────────────────────────────────────
  Return on Investment (Gross / Budget)
──────────────────────────────────────────────────────────────────────────────────
                       Toy Story (1995) |██████████████████████████████| 32.08 (54.8%)
           Sense and Sensibility (1995) |███████                       |  8.41 (14.4%)
  Ace Ventura: When Nature Calls (1995) |██████                        |  7.08 (12.1%)
                       GoldenEye (1995) |█████                         |  5.87 (10.0%)
               Waiting to Exhale (1995) |████                          |  5.09 ( 8.7%)
──────────────────────────────────────────────────────────────────────────────────
  Total: 58.53  |  Unique: 5

1.52 ms ± 54.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


> *Toy Story* выигрывает третий раунд подряд: ROI **32×**. Для сравнения, следующий в списке, *Sense and Sensibility*, показывает «всего» 8,4×. Так рождаются легенды о том, что анимация — одна из самых недооценённых ставок в истории кино.

---

# Итоги

Кино — молодое искусство, но оно унаследовало главную функцию рассказа историй: помогать людям понимать себя и друг друга. Разница в том, что теперь у этого процесса есть побочный продукт — данные, по которым можно проследить, чего хочет массовый зритель в конкретный момент времени.

1. **Динамика релизов и жанры**
   * Пик кинопроизводства в датасете пришелся на 2000–2009 годы (до 311 релизов в год).
   * Базовые жанры каталога — Drama (4361) и Comedy (3756): два самых универсальных эмоциональных запроса, радость и рефлексия. Sci-Fi и Fantasy стабильно входят в топ-10 на протяжении всей истории каталога — воображение будущего оказывается такой же постоянной потребностью, как и смех.

2. **Пользовательские теги**
   * Самый частый тег — служебный `In Netflix queue` (131 вызов). Среди содержательных лидируют `atmospheric`, `thought-provoking` и `surreal`.
   * Теги про «robots» и андроидов подтверждают устойчивый интерес к историям о будущем, уже увиденный в жанровой статистике.

3. **Оценки и поведение аудитории**
   * Распределение смещено в плюс: чаще всего ставят 4.0 (26.8k) и 3.0 (20k) — аудитория скорее снисходительна, чем строга (лишь 8% зрителей — стабильно жёсткие критики).
   * Наибольший раскол мнений вызывают не блокбастеры, а авторское кино (*«Иваново детство»*, *«Фанни и Александр»*) — там, где сталкиваются разные представления о том, каким должно быть кино.
   * Фильмы 40–50-х годов (*Sunset Blvd.*, *A Streetcar Named Desire*) стабильно удерживают средний балл ≥ 4.3 при небольшом числе голосов — признание есть, просто не массовое.

4. **Экономика и окупаемость (TMDb)**
   * Большой бюджет не гарантирует успеха: *Cutthroat Island* ($102M бюджета) осталась убыточной.
   * Рекордную окупаемость показывают не самые дорогие, а самые оригинальные проекты: *Toy Story* достигла ROI ≈ 32× ($932M чистой прибыли при бюджете $30M).

**Жанры, оценки и деньги** — это четыре разных способа зафиксировать один и тот же импульс: что людям на самом деле нужно от историй. Человечество лишь недавно смогли так масштабно анализировать эти процессы. Последнее время мы плакали и смеялись, мечтали, наблюдали за чужими судьбами. Быть может, не за горами время, когда нам захочется новых историй.